### DISCUSSION TITLE
Localization in new admin console (and in admin REST API)

### DISCUSSION THREAD

**Discussion by mposolda** [keycloak/keycloak#16882] [Feedback]
With the new admin console, there is a question how we should approach localization. There are 2 approaches:

1) Let the admin REST API to return localized values directly in the JSON objects. So for example ServerInfoEndpoint will return in the `providers` field and their options with already translated `label` and `helpText`.

2) Let the admin REST API to return "keys" in the `label` and `helpText`. And then the admin console itself will be responsible for translating it.

We discussed with @jonkoops and he proposed approach (1) and I agree that it can be better. But one possible concern is, if admin REST API has all the information it needs to detect correct language? It has access to current user, which is most important. However I see the potential limitation that admin REST API doesn't have access to `KEYCLOAK_LOCALE` cookie or to `Accept-Language` HTTP header, which was set in the HTTP request to the UI. And those are also used to detect locale. So I guess that new admin console may need to send `Accept-Language` header in the HTTP requests to the admin REST API based on the locale from `KEYCLOAK_LOCALE` and/or `Accept-Language` headers sent in the HTTP request to UI. It might be good to have some "generic" filter/interceptor on admin console side and automatically send it to all REST requests. Will it be technically possible?

For the old admin console, we used approach (2), but with a bit cumbersome format returned in admin REST API, which was specific to Angular1 needs  - https://github.com/keycloak/keycloak/blob/20.0.0/services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpReviewProfileAuthenticatorFactory.java#L104 . This was not very elegant IMO and it will be nice if stuff returned from REST API does not need to be "tuned" according to the UI needs.

Regardless of approach (1) or (2), we can hopefully stick to have `label` and `helpText` keys to be in the "generic" format at the authenticators. But in case of approach (1), the endpoint itself (EG. ServerInfoEndpoint) will be responsible for translation of the label. This means that for individual authenticators are not responsible for translation and can still use cleaner approach with something like `myOption.setLabel("myOptionLocalizationKey")` instead of `myOption.setLabel(pleaseTranslateMe("myOptionLocalizationKey")`.

**Comment by mposolda** [comment]
It looks that when you have set in your browser the preferred language "cs" and you send request to Keycloak admin console, it has `Accept-Language` header correctly set with `cs`. I see it is propagated in the admin REST API request as well. So looks that we should be good already as we can consume directly the `Accept-Language` header in the REST requests. So even stronger argument for (1) and handle localization on server and REST API returned already translated values.

**Reply by jonkoops** [reply]
We could also set this header manually if needed, so it's not an issue either way.

**Comment by jonkoops** [comment]
So my opinion on this is pretty clear, all labels should be translated on an API level. For a couple of reasons:

1. In the current the API is inconsistent, sometimes keys are returned, sometimes translated strings (we need to pick one).
2. It makes the API hard to consume for other involved parties as they  now have to do additional mapping as well.
3. It makes it hard to tell which labels are actually used on the front-end, making it hard to remove unused labels confidently.

**Comment by ssilvert** [comment]
> So my opinion on this is pretty clear, all labels should be translated on an API level. For a couple of reasons:

I completely disagree.  Human translation is part of the human interface (UI).  In other words, the renderer is responsible for rendering.  And the server can't predict how the UI will want to render something, even when it comes to translations.

The problem that @mposolda brings up is just one example.  The server can't be sure of the locale.  So because the server needs to be certain of the locale to do a translation, it has to go through different machinations to find out.  And no matter how you try to find out, there will be limitations that constantly need to be addressed.  For instance, what if the app has its own out-of-band method for setting the locale?  You would have to account for that in the REST API.  You would need some explicit way of setting the locale with every REST call.

Here is another example of why rendering has to be up to the renderer.  Let's say that you want render based on layout or screen size.  Maybe you want to localize to an abbreviated translation when the screen is small.  

If translation is done on the server side then this will be impossible because the server always returns the long version of the translation.  So just like with finding the locale, you would have to communicate something about the client to the REST API.  In this case, you would need to send the screen size back to the server.  

Again, rendering has to be left up to the renderer.  Doing it inside each REST API call is architecturally flawed.

> 
> 1. In the current the API is inconsistent, sometimes keys are returned, sometimes translated strings (we need to pick one).

We don't need to pick one at all.  The API sends back a label.  That label should usually be returned as a key.  But if the back end developer is lazy and sends back plain text, the key will not be found and the plain text will be rendered instead.  That's how all the localization libraries work.  You pass a key.  If key is not found you just get back your param unchanged.

`translate("username")` returns `User name`  <--key was found
`translate("User name")` returns `User name`  <--- key not found

> 2. It makes the API hard to consume for other involved parties as they now have to do additional mapping as well.

The mapping has to be done somewhere.  The question is where does it belong from an architectural point of view?  I contend that the REST API is only responsible for returning data and metadata.  The UI is responsible for rendering the data based on its metadata.  In this case, the metadata is a localization key.  Final rendering should only be done by the renderer.

> 3. It makes it hard to tell which labels are actually used on the front-end, making it hard to remove unused labels confidently.

I don't see how server-side rendering solves that.  No matter where the rendering is done, there will be keys and message bundles.  It will always be a challenge to keep up with unused keys.  And keys can be defined by the Keycloak admin or from the developer of a provider.  Rendering on the server side doesn't change the fact that the keys have to be managed somewhere (probably multiple "somwheres").

**Comment by edewit** [comment]
I'm plus one on 1. we want an API that is usable for many clients, making it return cryptic keys that need another lookup does the opposite. On your point @ssilvert that the API should return data and isn't responsible for the rendering I do agree, but in the case of keycloak this is not true. At the moment the API does do "rendering" as it returns the fields that make up the form.  

> The server can't be sure of the locale. So because the server needs to be certain of the locale to do a translation, it has to go through different machinations to find out

We have the browser that sends that with every request and we have a logged in user, I don't think this is hard and not something new that needs to be introduced.

> We don't need to pick one at all. The API sends back a label.

If we decide to do that we need to change a lot of code, because atm it is sending back text in English. And sure we can still translate them and pretend that "User name" is a key but if the text is `If 'Remove none' is chosen, the user will keep all existing OTP configurations (legacy behavior).` then using that as key is no longer an option.

> The mapping has to be done somewhere. 

Yes so why not do it in one place once and not in `x` places. From an architectural point of view, doing it once is better then spreading it out to every client using the API

**Comment by pedroigor** [comment]
I think we have a draw then ... I'm with @ssilvert.

```
At the moment the API does do "rendering" as it returns the fields that make up the form.
```
The UI does the rendering backed by metadata provided by the server. In the case of localization, the server gives you only the key (metadata). I agree with @ssilvert that it is a better design with a clear separation of responsibilities (the main point of discussion) and more flexibility on the client side (less chance to introduce server-side changes for UI-related behavior).

For instance, in the case of https://github.com/keycloak/keycloak/issues/22507 users should make sure the keys used in the metadata are translatable from the client (the admin console).

However, it is a very bad experience having to update not only the Admin UI messages but also the messages from themes in order to render the text when at the login or registration pages. But once those pages are React-based, we could have a single (client-side) source/mechanism for localization covering both admin and current templates, right?

**Comment by ssilvert** [comment]
The UI team had a meeting about this.  I'm not sure anyone's mind was changed, but one thing did come up that I think is hard to refute.  That is how to handle parameterized messages, or [what i18next calls "Interpolation"](https://www.i18next.com/translation-function/interpolation).

We use parameterized messages in the admin console right now.  For instance, we have this in the message bundle:
`"leaveGroupConfirmDialog_one": "Are you sure you want to remove {{username}} from the group {{groupname}}?"`

To implement this on the server side you would have to send the params to the server for rendering.  This makes no sense because the selected `username` and `groupname` are purely client side UI state.

**Reply by pedroigor** [reply]
What I don't see is why we need to implement that message in particular on the server side. For me, it is clearly a client-side message and by dealing with it on the server side we are making the server aware of the UI/UX of the client. Other clients might want a different message for the same thing, right?

**Reply by ssilvert** [reply]
Right.  That's exactly my point and I think it underscores how separation of concerns can keep you out of trouble.

**Reply by jonkoops** [reply]
I have no idea why this is being used as an example, as this is clearly a message that belongs specifically to the client.

**Reply by ssilvert** [reply]
Yeah, I wasn't really addressing the part about `Other clients might want a different message for the same thing`.  Just that it's clearly a client-side message containing client state and it only makes sense to render it on the client.

To do this on the server, you would have to send client state to the server, let the server do the rendering, and get back a response.  You would have to do that for every interpolated message for every user every time you want to display it.

**Reply by jonkoops** [reply]
This is true, but I don't really see an issue with that, it's a pretty common element for localization. And we're already doing this historically in some templates I believe.

**Reply by ssilvert** [reply]
I don't understand your comment.  Server-side rendering of client state is clearly an issue.

**Reply by pedroigor** [reply]
@jonkoops Yeah, in templates, the server is resolving messages but just because the server is serving the client. The server is not really dealing with any client state when resolving messages.

It is a good discussion. We are probably missing updates to i18n support in Keycloak and if what you are suggesting is a common element for localization, I'm interested to learn more about how it works and the motivation for doing it.

**Reply by jonkoops** [reply]
> I don't understand your comment. Server-side rendering of client state is clearly an issue.

We're talking past each other. It's perfectly fine to do interpolation on the server where needed. Client-side messages should be on the client. So we're in agreement.

**Reply by jonkoops** [reply]
> Yeah, in templates, the server is resolving messages but just because the server is serving the client. The server is not really dealing with any client state when resolving messages.

Exactly, and I am not propagating the idea that this should happen. What I am saying is that the fields defined for User Profile (and other places) are user-defined content. This has nothing to do with our existing systems for localizing messages such as the aforementioned one from the client.

> and if what you are suggesting is a common element for localization, I'm interested to learn more about how it works and the motivation for doing it.

Yes! There is a commonality here, we have many places where users might want to show different strings depending on what locale is preferred by the user, for example; the terms of service, the names and descriptions of clients, the display name of a realm, etc.

For example, let's take the `name` of a client. Currently as a user, I have to provide an expression as a value if I want to localize it. For example `${my_client}`, then I have to navigate to the realm settings, and add a message with the key `my_client` for each locale I am want to support.

The first issue with this approach is the poor user-experience. If a user is expected to provide localized strings for the client name, the most intuitive option would be to allow setting those values in the same form. For example, by having a drop-down in the form that allows you to select the locale. Something that is very common in the CMS space for example:

![A screenshot of a CMS that supports multiple locales.](https://github.com/keycloak/keycloak/assets/695720/4e7c3ce4-bb52-4959-b615-5961b4054487)

It is also a problem from a front-end perspective, as we have no way of knowing which labels will be used when. For example, if we were to localize the name of the client we must do the following:

1. Load information about the client, get the value of the name.
2. Evaluate the name as a bespoke (to my knowledge) expression and figure out which keys are used. (nothing prevents a user from doing `${someLabel} BLA BLA ${someOtherLabel}`).
3. Request ALL user-defined labels, because we cannot filter these (as far as I am aware). Note, there are users which have thousands of these labels (they shouldn't, but they do).
4. Load all the labels into memory, then pick the labels that match the template and apply them.
5. Show the newly formatted name of the client.

From the users' perspective and the technical perspective of the client, this creates a lot of overhead. And it doesn't have to be this way if we could handle the localization of fields in data in the API itself.

If we continue to use the example of a client, we could have the following data structure (as an opt-in, likely with a query parameter or something) for the JSON data going over the line:

```json
{
  "title": {
    "en-US": "Swift Banking",
    "nl-NL": "Swift Bankieren"
  },
  "description": {
    "en-US": "A banking app for iOS and Android!",
    "nl-NL": "Een bankieren app voor iOS en Android!"
  }
}
```

Here instead of getting a field with a single value for English strings, or the strange expression for mapping to message bundles, we get another level with keys that match the locales that are configured as supported in the realm. Each of these keys then contains the translated string that matches that locale.

This would allow us to easily make these forms on the front-end, and removes the technical jumping through hoops to get the labels mapped correctly from a different source (such as the message bundles).

As for the non-admin side of things. If I were to register an account and set my preferred language to Dutch, and then request information about this client, the following information could be retrieved directly:

```json
{
  "title": "Swift Bankieren",
  "description": "Een bankieren app voor iOS en Android!"
}
```

This is extremely useful for applications such as the Account Console, as this completely removes the work for us to display this data accurately. And not just for us, but also for any other consumer of this API such as plugins, or custom applications that our users are running.

So in the end, what I am trying to say is that we should make a distinction between translated **content**, and translated **messages**. As doing so would result in a much more intuitive and consistent user-experience and developer-experience.

**Reply by jonkoops** [reply]
@xianli123 from a UX perspective this might also be interesting for you. Specifically how we could better handle translating user-generated content in the Admin Console.

**Reply by xianli123** [reply]
@jonkoops Thanks. I will take a look. 🤓

**Comment by edewit** [comment]
What me and Jon are talking about are the cases where the server side dictates what needs to be shown. For instance the dynamic fields that can be added by implementing some SPI or the user profile. It seems to me that when the server is sending back what fields and text to show it should do this in the language that the user wants, not in English and not by adding keys. Because this adds more complexity to actually add the field and we have to have define things at different locations and think about what overrides what, keeping the name of the field in the same place as the name makes more sense to me.

Another problem is the dynamic nature of the realm localisation settings, this might be a edge case, but it makes doing any kind of static analysis impossible. On top of that, like Jon mentioned, it's not very user friendly as you use to remember what the key needs to be.

What we are not talking about are the cases where the UI is creating the entire UI, in that case it makes total sense that the localisation is done on the client side.

IMO what we need to get rid off is the database managed locale stings and the English texts in the SPI's, but instead have the localised stings for them in the endpoint.

**Reply by ssilvert** [reply]
Ah, so we are only talking about the dynamic parts of the UI?  That does change the perspective a bit.  Need to think through it.

1. Are you proposing that localization of the display name should be specified along with the attribute on 
`Realm settings->User profile->Edit attribute`?

2. Then server sends display name to the client as a localized string? 

Number 1 makes a lot of sense from a UXD perspective.

I'm still not sure about number 2.  That would make at least four ways to do localization in Keycloak.  I think everyone agrees that we need to clean up this mess.  But maybe that's a separate issue.

So an advantage I see of number 2 is that you only send the attributes you need.  I believe you guys already mentioned this.  And that would be important if you really do have thousands of attributes.  

The downside is that you would not be allowed to use parameterized strings that depend on client side state.  Maybe that's OK for now.  But if we move more and more toward dynamic UI then we will miss the ability to use interpolation.

And just to add more fuel to the fire, I have just realized that for any User Profile attribute rendered as a dropdown, we need a way to localize the dropdown values.  Is this there already?

**Reply by edewit** [reply]
> I'm still not sure about number 2. That would make at least four ways to do localization in Keycloak. I think everyone agrees that we need to clean up this mess. But maybe that's a separate issue.

Right we might need an separate issue for that, but what Jon and I are suggesting is not adding yet another way to add localisation but to remove two. Let's get rid of the 'dynamic' realm specific keys and the server side overrides and merge it into one. 

> The downside is that you would not be allowed to use parameterized strings that depend on client side state. Maybe that's OK for now. But if we move more and more toward dynamic UI then we will miss the ability to use interpolation.

We don't do this and we don't need to, so this is not really a valid argument. 

> And just to add more fuel to the fire, I have just realized that for any User Profile attribute rendered as a dropdown, we need a way to localize the dropdown values. Is this there already?

Yes, and if you think about how many keys you have to define for one field: the name of the field the select items the help text, the group they are in, and that times all the locales you support
